### ISSUE TITLE
About Weibo Dataset

### ISSUE THREAD

**Issue by Sammir-z** [yliuaa/MIMoE-FND#6]
Hello, the author. After I processed the Weibo data set, the training set was only 7411, and the test set was only 1885. Then the effect of testing was only 0.8832. I don't know whether it was the wrong amount of data I processed or other reasons.

![Image](https://github.com/user-attachments/assets/de6b8728-dfdc-4408-b5df-d448ccd91842)

**Comment by Sammir-z** [comment]
When I was processing the dataset, because access to the microblog gallery was forbidden, I downloaded it through an API of Baidu. If I followed the original processing code, most of the Weibo dataset would not be available. When you are training and testing, is it convenient for you to talk about the Weibo dataset?

**Comment by yliuaa** [comment]
I can share my processed weibo dataset. Please leave an email. I can’t really redistribute the original weibo though.

**Comment by Sammir-z** [comment]
OK, thank you very much for your work. My email is zhanggb928@163.com. I have Weibo's original data, but I don't know whether the processing method is quite right.

**Comment by Sammir-z** [comment]
Here's my code for handling Weibo

``` Python
def weibo_text_to_excel(is_train="both",conduct_download=True):
    is_train_list = ["train","test"]
    num_too_small = 0
    # weibo_train_id, weibo_val_id, weibo_test_id = None ,None ,None
    num_length, num_cv_error, num_no_valid = 0, 0, 0
    train_list, val_list, test_list = [], [], []
    for is_train in is_train_list:
        non_rumor_text = '/gemini/code/MIMoE-FND-main/Dataset/weibo/weibo_dataset/tweets/{}_nonrumor.txt'.format(is_train)
        rumor_text = '/gemini/code/MIMoE-FND-main/Dataset/weibo/weibo_dataset/tweets/{}_rumor.txt'.format(is_train)
        text_lists = [rumor_text, non_rumor_text]
        results = []
        for i in tqdm(range(len(text_lists))):
            text_list = text_lists[i]
            f = open(text_list,encoding='UTF-8')  # 返回一个文件对象
            line = f.readline()  # 调用文件的 readline()方法
            line_num,results_dict = 0, {}
            while line:  # 如果每行是json的字符串形式  json.loads(line) # line_json["filename"])
                if line_num%3==0: # 无用信息
                    results_dict['source'] = line
                elif line_num%3==1: # img
                    # imgs = line.split('|')
                    # imgs = [img[img.rfind('/')+1] for img in imgs]
                    results_dict['images'] = line #imgs
                else: # content
                    results_dict['content'] = line
                    results_dict['label'] = i
                    results.append(results_dict)
                    results_dict = {}
                line = f.readline()
                line_num+=1
            f.close()

        df = pd.DataFrame(results)
        filename = '/gemini/code/MIMoE-FND-main/Dataset/weibo/weibo_dataset/origin_do_not_modify/{}_datasets_origin.xlsx'.format(is_train)
        df.to_excel(filename)
        print(f"File saved at {filename}")

        folder1, folder2 = "/gemini/code/MIMoE-FND-main/Dataset/weibo/weibo_dataset/nonrumor_images", "/gemini/code/MIMoE-FND-main/Dataset/weibo/weibo_dataset/rumor_images"
        images1, images2 = set(os.listdir(folder1)), set(os.listdir(folder2))
        images_set = images1.union(images2)
        news_xlsxs = []
        if is_train=="train" or is_train=='both':
            news_xlsxs.append('/gemini/code/MIMoE-FND-main/Dataset/weibo/weibo_dataset/origin_do_not_modify/train_datasets_origin.xlsx')
        if is_train == "test" or is_train == 'both':
            news_xlsxs.append('/gemini/code/MIMoE-FND-main/Dataset/weibo/weibo_dataset/origin_do_not_modify/test_datasets_origin.xlsx')

        for news_xlsx in news_xlsxs:
            wb = openpyxl.load_workbook(news_xlsx)
            sheetnames = wb.sheetnames
            sheet = wb[sheetnames[0]]
            rows = sheet.max_row
            for i in tqdm(range(2, rows + 1)):
                source = sheet['B' + str(i)].value
                title, full_valid_name = '', []
                content = sheet['D' + str(i)].value
                if content is None or len(content) <=10:
                    print("Found length not satisfy {}".format(content))
                    num_length += 1
                    continue
                image_names = sheet['C' + str(i)].value
                label = int(sheet['E' + str(i)].value)
                if image_names is not None and len(image_names) != 0:
                    image_name_list = image_names.split('|')[:-1]
                    for image_name in image_name_list:
                        full_image_name = image_name.lstrip(' ').rstrip(' ')  # http url
                        short_name = full_image_name[full_image_name.rfind('/') + 1:]
                        ## if not found, we tempt to download... if cannot download, we skip...
                        if short_name not in images_set:
                            if conduct_download:
                                try:
                                    html_path = "https://image.baidu.com/search/down?url=" # 因为原本微博图库的链接失效，只有转换链接了
                                    full_image_name=html_path+full_image_name
                                    wget.download(full_image_name, "/gemini/code/MIMoE-FND-main/Dataset/weibo/weibo_dataset/"+("rumor_images/" if label == 0 else "nonrumor_images/")  + short_name) # 这里我没有使用templ 文件报错，因为好像如果使用就会报错
                                    # image_name = image_name + "rumor_images/" + short_name + "|"
                                    # print("Download ok. {}".format(full_image_name))
                                    images_set.add(short_name)
                                except Exception:
                                    print("Download error. {}".format(full_image_name))
                            else:
                                print("Do not download. {}".format(full_image_name))

                        if short_name in images_set:
                            new_image = ("rumor_images/" if label == 0 else "nonrumor_images/") + short_name
                            is_found_image = cv2.imread("/gemini/code/MIMoE-FND-main/Dataset/weibo/weibo_dataset/"+new_image)
                            if is_found_image is not None:
                                image_width, image_height = is_found_image.shape[0], is_found_image.shape[1]
                                ## CONDUCT FILTERING OUT INVALID NEWS IF NOT NO_FILTER_OUT
                                ratio = float(image_width) / image_height
                                # if ratio>3 or ratio<0.33 or image_width < 100 or image_height < 100:
                                #     news_type = "text"
                                #     print("Size {} too small {} skip..".format(is_found_image.shape,new_image))
                                #     num_too_small += 1
                                # else:
                                full_valid_name.append(new_image)
                if len(full_valid_name)!=0:
                    full_valid_name = '|'.join(full_valid_name)
                    record = {}
                    record['title'], record['label'], record['source'], record['content'], record['image'] \
                        = title, label, source, content, full_valid_name
                    ## send ''category'' to the data loader, so that it can replace some of the useless modalities into placeholders.
                    # record['category'] =
                    query_key = source.split('|')[0]
                    if "train" in news_xlsx:
                        train_list.append(record)
                    else:
                        test_list.append(record)
                else:
                    print("No valid image found {}".format(content))
                    num_no_valid += 1

    # if "train" in is_train:
    filename = '/gemini/code/MIMoE-FND-main/Dataset/weibo/weibo_dataset/origin_do_not_modify/train_datasets_WWW_new.xlsx'
    df = pd.DataFrame(train_list)
    df.to_excel(filename)
    print(f"File saved at {filename}")
    filename = '/gemini/code/MIMoE-FND-main/Dataset/weibo/weibo_dataset/origin_do_not_modify/test_datasets_WWW_new.xlsx'
    df = pd.DataFrame(test_list)
    df.to_excel(filename)
    print(f"File saved at {filename}")
    print("length {} CV_Error {} No_valid {}".format(num_length,num_cv_error, num_no_valid))
    print("Train_set {} Test_set {} Val_set {}".format(len(train_list),len(test_list),len(val_list)))
    print(f"Size not valid {num_too_small}")

if __name__ == '__main__':
    weibo_text_to_excel('both')
```
